Adding Python environments with README instructions; starting to extract JIRA data from a xml file

IanSinColorado opened the pull request:
Check out the instructions in the README and see if the code segments work for reproducing the environment(s)

IanSinColorado committed:
Added conda environment through .yml file

IanSinColorado committed:
Updated README.md to include a list of instructions on how to set up the conda environment with the provided environment.yml; also added a dev environment that includes VIM

IanSinColorado committed:
Very minor README change

IanSinColorado committed:
Added the beautifulsoup4 library to environment; beautifulsoup4 is a well documented and frequently used XML, HTML, and text parser that I have used in the past

IanSinColorado committed:
Minor update to readme

MattF-NSIDC commented:
Hey, @IanSinColorado ! Hope you don't mind my comments. The new repo appeared in my notifications and I decided to take a peek :)

In my experience the easiest way to deal with Conda environments is hand-maintaining an `environment.yml` which contains only your top-level dependencies (including both dev and runtime dependencies) and no sub-dependencies (i.e. include `beautifulsoup4`, but not `urllib3`).

`conda env export` is producing a record of your exact environment at that point in time, which may not be installable on other platforms and may even stop working next month on the same computer. If you instead hand-maintain an `environment.yml` which uses [compatible release](https://peps.python.org/pep-0440/#compatible-release) specification (e.g. `beautifulsoup4 ~=4.12`) and contains only things you directly `import`, then environment maintenance will be much easier over time. If you take this approach, I recommend avoiding `conda install <some-package>`, because that won't update your `environment.yml`. Instead, you could add `some-package` to your `environment.yml`, then do `conda env update` to apply the change to your environment.

An environment file with exact pins like you have is really useful if you want to build a docker image or run CI, but if you're not doing those things you may not need one at all. On NSIDC projects, we typically call those "lock files" following the nomenclature from NPM's `package-lock.json`, and often save them as `environment-lock.yml`. Some things to be aware of when using `conda env export`:

* the YAML outputs are platform-specific, and if you're using a Mac, that lock file may not work in GitHub Actions' Linux runners or on other developers' computers. For that, there's a tool called `conda-lock` which can create multi-platform lock files.
* the YAML outputs are specific to your personal computer; the last line of the YAML is `prefix: <some-path-on-your-computer>` and can (should?) be omitted. A one-liner: `conda env export | grep -v "^prefix:" > environment-lock.yml`
* the YAML outputs will contain your channel configuration from your `~/.condarc` file, which may not match the project's needs

[Here's](https://github.com/nsidc/sipn-reanalysis-plots/blob/main/environment.yml) an example of an `environment.yml` that's maintained with the above practices. The same repo also has an [`environment-lock.yml` file](https://github.com/nsidc/sipn-reanalysis-plots/blob/main/environment-lock.yml) which we produce using `conda env export` on a Linux machine so it can be built in GitHub Actions.

I'd also recommend against including your individual development dependencies (e.g. `vim`, `vscode`) in a project conda environment, and instead install them with Homebrew/Apt/whatever, as these may conflict with installed tools on other people's computers! Shared dev dependencies like `pytest` or `pre-commit` are more appropriate for `environment.yml`.

IanSinColorado committed:
manually made environment.yml to include requests and beautifulsoup4 and lxml; need to remove lxml; added JIRAloading.py to start testing beautifulsoup4; managed to get the environment to work and get the titles of 88 JIRA issues from the XML; need to learn how to use the GitHub issues API to start loading data

IanSinColorado commented:
I just added a commit with a manual environment.yml file. I tested it by starting to parse through the test JIRA issues XML file

MattF-NSIDC commented:
Looks great! Good call on writing up a quick test :sunglasses: I have a couple small suggestions:

* The `anaconda` channel is OK, but it's not worth using compared to the `conda-forge` channel. Packages on `anaconda` are built in private by ContinuumIO with little community input or control, while `conda-forge` packages are community maintained and built in the open from "feedstocks", which are GitHub repos containing packaging "recipes" only. I love this because when I need a new version that's not available, ContinuumIO won't take my call about updating the `anaconda` channel, but I can easily [open a feedstock PR](https://github.com/conda-forge/quarto-feedstock/pull/6) to get the version I need released onto `conda-forge`. I suggest replacing `anaconda` with `conda-forge` and adding the channel `nodefaults` to the end of the list. This prevents conda from using the `anaconda` channel implicitly, which can be annoying. I'd even suggest creating or updating a `~/.condarc` that sets the channels for your user to always be `conda-forge` and `nodefaults`.
* [beautifulsoup4](https://anaconda.org/conda-forge/beautifulsoup4) and [lxml](https://anaconda.org/conda-forge/lxml) are both available on `conda-forge`, so those can be pulled out of the `pip` section and into the main dependency list; you don't need the `pip` section at all :)

IanSinColorado committed:
edited file to no longer use deleted file; file was not meant to be
public

IanSinColorado committed:
added nodefaults as final channel to environment.yml; this ensures the anaconda channel is never used

IanSinColorado committed:
Update README.md

Removed notes about the conda environment.yml file as it is now being maintained manually rather than through exporting the current environment

IanSinColorado committed:
Deleting swap file

IanSinColorado committed:
adding a function to get titles from xml

IanSinColorado committed:
making functions to load the information from JIRA issues; got titles,
links, and descriptions; need to get rid of the xml tags in the
descriptions (with regex?)

IanSinColorado committed:
Create .gitignore

Used GitHub's default Python template, with the addition of swap files at the very bottom

IanSinColorado committed:
Needed lxml parsing library to read .xml from ubuntu terminal; added
functions to remove the tags from descriptions (except for annoying a
tags) and extract / see how many lists there are in the description (not
a checklist)

IanSinColorado committed:
rewrote loading functions to use beautifulsoup instead of failing to use
regex; still need to remove the tags from description using .contents

IanSinColorado committed:
added function for getting number of tags in an issue; trying to get a
description without any tags

IanSinColorado committed:
adjusted the description loading function to extract the text from the
description tags; still need to remove the a tags that are there for
links

IanSinColorado committed:
created three functions; 1.) to extract the descriptions as raw html
for getting more data later, 2.) extracting all of the a tags from each
issue's description, and 3.) to extract the description as raw text
without the tag

IanSinColorado commented:
I have completed the issue of creating a reproducible conda environment. There are instructions in the README if you want to test it out. I think we need to merge this pull request and start a new branch/pull request, but in the meantime, I have started to load JIRA issue data from the jiraIssuesCryo.xml file. 
So far, I have functions to extract the title, JIRA issue link, and description (raw and pretty). I still need to extract the labels, status, and assignees from the XML file. Finally, I need to check if any other JIRA issue fields can be migrated to GitHub.

IanSinColorado committed:
Created second python script for uploading the data to GitHub; the
script works and is revealing errors in the descriptions; also change
the JIRAloading.py file to include a class to use the
functions in GitHubAPI.py more efficiently

IanSinColorado committed:
Added a function that extracts the labels and status from the issues

IanSinColorado committed:
removed the regex parsing for lists as the GitHub api can
comprehend HTML; cleaned up JIRAloading.py; added checklists to GitHub,
but unsure how successful I was because the test data only had one
checklist in it; added an update_issue() function and also adjusted the
upload_issue() function to also return the issue number

jwfuller commented on README.md:
```diff
@@ -24,6 +24,23 @@ nsidc@nsidc.org for more information.
+This section will serve as a guide to set up the proper environment with the required libraries.
+
+1. Clone this repository into a directory of your choosing
+2. Open the Anaconda Prompt (do a search on your computer for 'Anaconda Prompt' or open any terminal that can use the `conda` keyword; check using `conda --version` to see if a version pops up)
```
```suggestion
2. Verify to have [conda installed](https://docs.conda.io/projects/conda/en/latest/user-guide/install/index.html): open a terminal and run `conda --version`. If a version number is returned, you have conda or miniconda installed.
```

jwfuller commented on GitHubAPI.py:
```diff
@@ -0,0 +1,130 @@
+REPO_OWNER = 'iansincolorado'
+REPO_NAME = 'pythonapis'
+
+# Only way of authenticating at the current moment is with GitHub PAT tokens, which are deleted from your GitHub account if it is found in a public repository
+PAT = ''
```
You could set the PAT as an environmental variable or prompt for it and throw an error if it is not set.

jwfuller commented on GitHubAPI.py:
```diff
@@ -0,0 +1,130 @@
+# Getting the issues loaded into the JIRALoader object
+issues.get_issue_data()
+
+REPO_OWNER = 'iansincolorado'
+REPO_NAME = 'pythonapis'
```
Could you prompt for these values?

jwfuller commented on README.md:
```diff
@@ -24,6 +24,23 @@ nsidc@nsidc.org for more information.
+- Use `conda remove -n JG-ETL --all` to remove the environment
+- Use `conda env update -f environment.yml` to update the environment after any changes to the environment.yml file
 
 ## Usage
```
Add the steps to run your scripts

jwfuller commented on JIRAloading.py:
```diff
@@ -0,0 +1,158 @@
+import requests
+import re
+
+with open('../testJIRA.xml', 'r') as f:
```
Prompt for file name

jwfuller commented on JIRAloading.py:
```diff
@@ -0,0 +1,158 @@
+
+# Getting all of the items, which are individual JIRA issues
+issues = soup.rss.channel.find_all('item')
+# print(type(issues))
+# print(len(issues))
+# print(issues[0])
```
Get rid of debugging statements

jwfuller commented on JIRAloading.py:
```diff
@@ -0,0 +1,158 @@
+
+
+
+    # This function compiles the data from the titles, links, descriptions, labels, and checklists into a zipped array of tuples, where each tuple is all of the data for one issue.
```
Extend comment to include why is it important that they are tuples

jwfuller commented on JIRAloading.py:
```diff
@@ -0,0 +1,158 @@
+
+
+
+# Testing
+test = JIRALoader('../testJIRA.xml')
+issues_data = test.get_issue_data()
+# for issue in issues_data:
+    # print(issue[1])
```
Is this debugging code?

jwfuller commented on GitHubAPI.py:
```diff
@@ -0,0 +1,130 @@
+    URL = f'https://api.github.com/repos/{REPO_OWNER}/{REPO_NAME}/issues'
+    headers = {
+        'Authorization': 'token ' + PAT
+        # 'accept': 'application/vnd.github+json'
```
Do you need this line in the future?

jwfuller commented on GitHubAPI.py:
```diff
@@ -0,0 +1,130 @@
+
+    r = None
+# ****************************************************
+# Use of try and exception handeling provided by GeeksForGeeks at https://www.geeksforgeeks.org/exception-handling-of-python-requests-module/
+    try:
+        r = requests.post(URL, data=json.dumps(data), headers=headers, timeout=5)
+        r.raise_for_status()
+    except requests.exceptions.HTTPError as errh:
+        print("HTTP Error")
+        print(errh.args[0])
+    except requests.exceptions.ReadTimeout as errrt:
+        print("Time out")
+    except requests.exceptions.ConnectionError as conerr:
+        print("Connection error")
+    except requests.exceptions.RequestException as errex:
+        print("Exception request")
+    print("Update Success")
+# ****************************************************
```
Could you reuse the code form above?

jwfuller reviewed (changes requested):
Awesome start!

IanSinColorado committed:
moved requests library request outside of functions for code reusability
and cleanup

IanSinColorado committed:
cleaned up JIRAloading.py to just be the class, with no more testing
lines; cleaned up GitHubAPI.py to have all testing code be at the bottom
of the file, rather than split across the top and bottom; need to add
inputs for the repo owner, repo name, JIRA xml file path, and PAT.

IanSinColorado committed:
Removed all testing from both files; added a run() function that is a
text-based UI for the user to use my tool; also added the
print_issue_titles() function to the JIRALoader class as a helper
function for run()

IanSinColorado committed:
renamed upload_issue() -> create_issue()

IanSinColorado committed:
updated the usage section to include instructions on PAT tokens and
how to use the tool through the command line
